File: 3.x.x/crates-utopia/messaging/src/adapter/sms/mock.rs
//! PHP `Utopia\Messaging\Adapter\SMS\Mock`.

use serde_json::json;

use super::{sms_response_from_status, TYPE};
use crate::adapter::{expect_sms, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::SMS;
use crate::response::ResponseData;

/// PHP `Adapter\SMS\Mock`.
#[derive(Debug)]
pub struct Mock {
    base: AdapterBase,
    user: String,
    secret: String,
    url: parking_lot::Mutex<String>,
}

impl Mock {
    /// PHP `__construct($user, $secret)`.
    #[must_use]
    pub fn new(user: impl Into<String>, secret: impl Into<String>) -> Self {
        Self {
            base: AdapterBase::default(),
            user: user.into(),
            secret: secret.into(),
            url: parking_lot::Mutex::new("http://request-catcher:5000/mock-sms".into()),
        }
    }

    /// PHP `getEndpoint`.
    #[must_use]
    pub fn get_endpoint(&self) -> String {
        self.url.lock().clone()
    }

    /// PHP `setEndpoint`.
    pub fn set_endpoint(&self, url: impl Into<String>) -> &Self {
        *self.url.lock() = url.into();
        self
    }

    fn process_sms(&self, message: &SMS) -> ResponseData {
        let url = self.get_endpoint();
        let result = self.request_default(
            "POST",
            &url,
            &[
                "Content-Type: application/json".into(),
                format!("X-Username: {}", self.user),
                format!("X-Key: {}", self.secret),
            ],
            Some(json!({
                "message": message.get_content(),
                "from": message.get_from(),
                "to": message.get_to().join(","),
            })),
        );
        sms_response_from_status(
            message,
            &result,
            |r| r.status_code == 200,
            |_| "Unknown Error.".into(),
        )
    }
}

impl Adapter for Mock {
    fn get_name(&self) -> &'static str {
        "Mock"
    }

    fn get_type(&self) -> &'static str {
        TYPE
    }

    fn get_message_type(&self) -> MessageKind {
        MessageKind::SMS
    }

    fn get_max_messages_per_request(&self) -> usize {
        1000
    }

    fn base(&self) -> &AdapterBase {
        &self.base
    }

    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(self.process_sms(expect_sms(message)?)))
    }
}

File: packages/messaging/src/Adapter/SMS/Mock.php
<?php

namespace Utopia\Messaging\Adapter\SMS;

use Utopia\Messaging\Adapter\SMS as SMSAdapter;
use Utopia\Messaging\Messages\SMS as SMSMessage;
use Utopia\Messaging\Response;

class Mock extends SMSAdapter
{
    protected const NAME = 'Mock';

    protected string $url = 'http://request-catcher:5000/mock-sms';

    /**
     * @param  string  $user User ID
     * @param  string  $secret User secret
     */
    public function __construct(
        private readonly string $user,
        private readonly string $secret,
    ) {
        parent::__construct();
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getEndpoint(): string
    {
        return $this->url;
    }

    public function setEndpoint(string $url): self
    {
        $this->url = $url;
        return $this;
    }

    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
     * {@inheritdoc}
     *
     * @throws \Exception
     */
    protected function process(SMSMessage $message): array
    {
        $response = new Response($this->getType());

        $response->setDeliveredTo(\count($message->getTo()));

        $result = $this->request(
            method: 'POST',
            url: $this->url,
            headers: [
                'Content-Type: application/json',
                "X-Username: {$this->user}",
                "X-Key: {$this->secret}",
            ],
            body: [
                'message' => $message->getContent(),
                'from' => $message->getFrom(),
                'to' => implode(',', $message->getTo()),
            ],
        );

        if ($result['statusCode'] === 200) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to);
            }
        } else {
            foreach ($message->getTo() as $to) {
                $response->addResult($to, 'Unknown Error.');
            }
        }

        return $response->toArray();
    }
}
